### PR TITLE
Add opIn and pointer-returning `BTreeMap.insert`.

### PR DESCRIPTION
**Pull request by nemanja-boric-sociomantic**
This adds a missing part of the API which allows accessing the stored values by pointer. Any change to these items will not affect the integrity of the map, since the map is ordered by keys, not by values.

### CODE REVIEW

File: src/ocean/util/container/btree/BTreeMap.d
Comment (mihails-strasuns-sociomantic) [inline src/ocean/util/container/btree/BTreeMap.d:179]: Meh, seeing more of the `out bool something` pattern I ask rest of core team to reconsider addition of `Optional` type into ocean :)

Comment (nemanja-boric-sociomantic) [reply src/ocean/util/container/btree/BTreeMap.d:179]: I think that's not a bad idea at all.

Comment (nemanja-boric-sociomantic) [reply src/ocean/util/container/btree/BTreeMap.d:179]: If you make a PR to get it included in ocean, I'll go and base this PR on top of that ;-)

File: src/ocean/util/container/btree/Implementation.d
Comment (david-eckardt-sociomantic) [inline src/ocean/util/container/btree/Implementation.d]:
```d
if (auto node = this.get(key, index))
    return &node.elements[index].value;
return null;
```

**Review by gavin-norman-sociomantic — COMMENTED** [review]
LGTM, pending the changes for `Optional`.

File: src/ocean/util/container/btree/BTreeMap.d
Comment (nemanja-boric-sociomantic) [reply src/ocean/util/container/btree/BTreeMap.d:179]: I'm now looking again into this and I think there's no need for `Optional` here :-(. Trick is that this method always returns a pointer, either to the old value or the new inserted value, passing the `bool` to the user to say if it added or removed the new element (it's akin to `getOrCreate` in Ocean's caches, really). Do you have proposal for the better API?

Comment (gavin-norman-sociomantic) [reply src/ocean/util/container/btree/BTreeMap.d:179]: Ha. Yeah, this function's API was misread. So you could add `out (p) { assert(p !is null); }`.

Comment (nemanja-boric-sociomantic) [reply src/ocean/util/container/btree/BTreeMap.d:179]: Yes, will add, along with the unittests.

**Comment by nemanja-boric-sociomantic** [comment]
Updated with an `out` contract and with unittests for the new features.

File: src/ocean/util/container/btree/BTreeMap.d
Comment (mihails-strasuns-sociomantic) [reply src/ocean/util/container/btree/BTreeMap.d:179]: :(

**Review by mihails-strasuns-sociomantic — APPROVED** [review]

**Pull request merged**
